File: src/class038/C6_SortStackWithRecursive.py
class Code06_SortStackWithRecursive:
    @staticmethod
    def sort(stack):
        """主函数，对栈进行排序"""
        depth = Code06_SortStackWithRecursive.get_depth(stack)
        while depth > 0:
            max_value = Code06_SortStackWithRecursive.get_max(stack, depth)
            count = Code06_SortStackWithRecursive.get_count(stack, depth, max_value)
            Code06_SortStackWithRecursive.sink_max(stack, depth, max_value, count)
            depth -= count

    @staticmethod
    def get_depth(stack):
        """计算栈的深度，不改变栈的内容"""
        if not stack:
            return 0
        val = stack.pop()
        depth = 1 + Code06_SortStackWithRecursive.get_depth(stack)
        stack.append(val)
        return depth

    @staticmethod
    def get_max(stack, depth):
        """寻找栈中深度为depth的最大值"""
        if depth == 0:
            return float('-inf')
        val = stack.pop()
        rest_max = Code06_SortStackWithRecursive.get_max(stack, depth - 1)
        max_value = max(val, rest_max)
        stack.append(val)
        return max_value

    @staticmethod
    def get_count(stack, depth, max_value):
        """计算最大值max在栈中出现的次数"""
        if depth == 0:
            return 0
        val = stack.pop()
        rest_count = Code06_SortStackWithRecursive.get_count(stack, depth - 1, max_value)
        count = rest_count + (1 if val == max_value else 0)
        stack.append(val)
        return count

    @staticmethod
    def sink_max(stack, depth, max_value, k):
        """将最大值沉到栈底"""
        if depth == 0:
            for _ in range(k):
                stack.append(max_value)
        else:
            val = stack.pop()
            Code06_SortStackWithRecursive.sink_max(stack, depth - 1, max_value, k)
            if val != max_value:
                stack.append(val)

# 测试程序
def main():
    test = [1, 5, 4, 5, 3, 2, 3, 1, 4, 2]
    Code06_SortStackWithRecursive.sort(test)
    while test:
        print(test.pop())

    # 随机测试
    import random
    print("测试开始")
    for _ in range(20000):
        stack = [random.randint(0, 19) for _ in range(20)]
        Code06_SortStackWithRecursive.sort(stack)
        sorted_correctly = all(stack[i] <= stack[i + 1] for i in range(len(stack) - 1))
        if not sorted_correctly:
            print("出错了!")
            break
    else:
        print("测试结束")

if __name__ == '__main__':
    main()

File: src/class038/C6_SortStackWithRecursive.cpp
#include <iostream>
#include <stack>
#include <algorithm> // For max

// 使用递归函数对栈进行排序
// 只使用栈的push、pop、empty三个方法以及递归函数，禁用其他容器和数组
class Code06_SortStackWithRecursive {
public:
    // 主函数，对栈进行排序
    static void sort(stack<int>& s) {
        int depth = getDepth(s);
        while (depth > 0) {
            int max = getMax(s, depth);
            int count = getCount(s, depth, max);
            sinkMax(s, depth, max, count);
            depth -= count;
        }
    }

private:
    // 计算栈的深度，不改变栈的内容
    static int getDepth(stack<int>& s) {
        if (s.empty()) {
            return 0;
        }
        int val = s.top();
        s.pop();
        int depth = 1 + getDepth(s);
        s.push(val);
        return depth;
    }

    // 寻找栈中深度为depth的最大值
    static int getMax(stack<int>& s, int depth) {
        if (depth == 0) {
            return INT_MIN;
        }
        int val = s.top();
        s.pop();
        int restMax = getMax(s, depth - 1);
        int maxVal = max(val, restMax);
        s.push(val);
        return maxVal;
    }

    // 计算最大值max在栈中出现的次数
    static int getCount(stack<int>& s, int depth, int max) {
        if (depth == 0) {
            return 0;
        }
        int val = s.top();
        s.pop();
        int restCount = getCount(s, depth - 1, max);
        int count = restCount + (val == max ? 1 : 0);
        s.push(val);
        return count;
    }

    // 将最大值沉到栈底
    static void sinkMax(stack<int>& s, int depth, int max, int k) {
        if (depth == 0) {
            for (int i = 0; i < k; i++) {
                s.push(max);
            }
        } else {
            int val = s.top();
            s.pop();
            sinkMax(s, depth - 1, max, k);
            if (val != max) {
                s.push(val);
            }
        }
    }
};

// 测试程序
int main() {
    stack<int> test;
    test.push(1);
    test.push(5);
    test.push(4);
    test.push(5);
    test.push(3);
    test.push(2);
    test.push(3);
    test.push(1);
    test.push(4);
    test.push(2);
    Code06_SortStackWithRecursive::sort(test);
    while (!test.empty()) {
        cout << test.top() << endl;
        test.pop();
    }

    // 随机测试
    cout << "测试开始" << endl;
    for (int i = 0; i < 20000; i++) {
        stack<int> stack;
        for (int j = 0; j < 20; j++) {
            stack.push(rand() % 20);
        }
        Code06_SortStackWithRecursive::sort(stack);
        bool sorted = true;
        int last = INT_MAX;
        while (!stack.empty()) {
            if (last < stack.top()) {
                sorted = false;
                break;
            }
            last = stack.top();
            stack.pop();
        }
        if (!sorted) {
            cout << "出错了!" << endl;
            break;
        }
    }
    cout << "测试结束" << endl;

    return 0;
}
